babel/_memoized.py

```diff
@@ -0,0 +1,36 @@
+# TODO: This can't live in .util until the circular import of
+#       core -> util -> localtime -> win32 -> core is resolved.
+
+
+class Memoized(type):
+    """
+    Metaclass for memoization based on __init__ args/kwargs.
+    """
+
+    def __new__(mcs, name, bases, dict):
+        if "_cache" not in dict:
+            dict["_cache"] = {}
+        if "_cache_lock" not in dict:
+            dict["_cache_lock"] = None
+        return type.__new__(mcs, name, bases, dict)
+
+    def __memoized_init__(cls, *args, **kwargs):
+        lock = cls._cache_lock
+        if hasattr(cls, "_get_memo_key"):
+            key = cls._get_memo_key(args, kwargs)
+        else:
+            key = (args or None, frozenset(kwargs.items()) or None)
```

Comment:
Maybe memoization could provide both a metaclass and a base class (using the metaclass), the latter providing a default _get_memo_key method? That would obviate the need for this conditional? (though not the `__new__` override)